mesonbuild/modules/python.py

```diff
@@ -297,6 +299,55 @@ def __init__(self, interpreter: 'Interpreter') -> None:
             'find_installation': self.find_installation,
         })
 
+    def _get_install_scripts(self) -> T.List[mesonlib.ExecutableSerialisation]:
+        backend = self.interpreter.backend
+        ret = []
+        installdata = backend.create_install_data()
+        py_files: T.Dict[str, T.List[T.Tuple[str, str]]] = collections.defaultdict(list)
+        optlevel = self.interpreter.environment.coredata.get_option(mesonlib.OptionKey('bytecompile', module='python'))
+        if optlevel == -1:
+            return ret
+
+        def should_append(f, isdir: bool = False):
+            return f.startswith(('{py_platlib}', '{py_purelib}')) and (f.endswith('.py') or isdir)
+
+        for t in installdata.targets:
+            if should_append(t.out_name):
+                py_files[t.subproject].append((t.out_name, os.path.join(installdata.prefix, t.outdir, os.path.basename(t.fname))))
+        for d in installdata.data:
+            if should_append(d.install_path_name):
+                py_files[d.subproject].append((d.install_path_name, os.path.join(installdata.prefix, d.install_path)))
+        for d in installdata.install_subdirs:
+            if should_append(d.install_path_name, True):
+                py_files[d.subproject].append((d.install_path_name, os.path.join(installdata.prefix, d.install_path, '')))
+        pycompile = os.path.join(self.interpreter.environment.get_scratch_dir(), 'pycompile.py')
+        if os.path.exists(pycompile):
+            os.remove(pycompile)
+        for i in self.installations.values():
+            if isinstance(i, PythonExternalProgram) and i.run_bytecompile[i.info['version']]:
```

Comment:
If I understand correctly, if we call `installation.install_sources()` it will compile all sources with that installation. Even sources installed with `another_installation.install_sources()`, or sources installed with `install_data(...,install_dir: another_installation.get_install_dir())`? Also if I only use `get_install_dir()` it won't compile any sources?

Not sure we have a way to track every installed file for which installation it is?

Reply:
> Also if I only use `get_install_dir()` it won't compile any sources?

Right, it probably makes sense to set `self.held_object.run_bytecompile` from inside the get_install_dir() method as well... seems safe to assume anyone using get_install_dir() but *not* install_sources() is actually using that dir to install other files. I probably just forgot to change this.

Reply:
What not removing `run_bytecompile` completely? If you did find_installation() chances are that you are going to install files. In worst case the manifest is just going to be empty and we can avoid adding installation script in that case.

Reply:
The number 1 real-world use of find_installation is:
```meson
py3 = import('python').find_installation('python3')

custom_target(
    output: 'foo.txt',
    command: [py3, files('makefoo.py')]
)
```

It baffles and frustrates me that people use find_installation when they really want find_program, but... they do. It's totally silly, because for example it means that on Debian they now depend on `apt install python3-distutils` for no reason whatsoever.

It seemed simpler to track whether to run bytecompile than to string-search every single installed file.

Reply:
Updated to set run_bytecompile for get_install_dir() too.